proto/substrait/algebra.proto

```diff
@@ -173,11 +173,13 @@ message JoinRel {
     JOIN_TYPE_OUTER = 2;
     JOIN_TYPE_LEFT = 3;
     JOIN_TYPE_RIGHT = 4;
-    JOIN_TYPE_SEMI = 5;
-    JOIN_TYPE_ANTI = 6;
+    JOIN_TYPE_SEMI = 5; // This is the left semi join.
+    JOIN_TYPE_ANTI = 6; // This is the left anti join.
     // This join is useful for nested sub-queries where we need exactly one record in output (or throw exception)
     // See Section 3.2 of https://15721.courses.cs.cmu.edu/spring2018/papers/16-optimizer2/hyperjoins-btw2017.pdf
     JOIN_TYPE_SINGLE = 7;
+    JOIN_TYPE_RIGHT_SEMI = 8;
+    JOIN_TYPE_RIGHT_ANTI = 9;
```

Comment:
Are there cases where a right semi join or a right anti join is fundamentally different than swapping the inputs to a left semi join or a left anti join?  Or are we just adding them for completeness' sake?  Do we need a right single join as well?

Reply:
Here we're adding it because it's present in all of the other join type enums.

Overall, the argument for having left and right is ease of roundtrip.  One can obviously switch the inputs but it doesn't appear to be a common practice despite its simplicity.

Reply:
Based on the description of the join (returning NULLs for one side) it seems like it should be added.